File: code/backend/java/adapter-rest/src/test/java/games/paths/adapters/rest/controller/dev/DevControllerTest.java
package games.paths.adapters.rest.controller.dev;

import games.paths.core.model.dev.CleanupResult;
import games.paths.core.port.dev.TestDataCleanupPort;

import org.junit.jupiter.api.Test;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

/**
 * Unit tests for {@link DevController}.
 */
class DevControllerTest {

    @Test
    void cleanup_shouldReturnCountsWhenDevEnabled() throws Exception {
        TestDataCleanupPort cleanupPort = mock(TestDataCleanupPort.class);
        when(cleanupPort.cleanupTestData()).thenReturn(new CleanupResult(5, 2));

        MockMvc mockMvc = MockMvcBuilders
                .standaloneSetup(new DevController(cleanupPort, true)).build();

        mockMvc.perform(post("/api/dev/cleanup"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.deletedGuests").value(5))
                .andExpect(jsonPath("$.deletedMatches").value(2));

        verify(cleanupPort).cleanupTestData();
    }

    @Test
    void cleanup_shouldReturn403WhenDevDisabled() throws Exception {
        TestDataCleanupPort cleanupPort = mock(TestDataCleanupPort.class);

        MockMvc mockMvc = MockMvcBuilders
                .standaloneSetup(new DevController(cleanupPort, false)).build();

        mockMvc.perform(post("/api/dev/cleanup"))
                .andExpect(status().isForbidden())
                .andExpect(jsonPath("$.error").value("DEV_ENDPOINTS_DISABLED"));

        // The cleanup use case must not run when dev endpoints are disabled.
        verify(cleanupPort, never()).cleanupTestData();
    }
}

File: code/backend/php/tests/Unit/Adapter/Rest/Dev/DevControllerTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Adapter\Rest\Dev;

use Games\Paths\Adapter\Rest\Dev\DevController;
use Games\Paths\Core\Domain\Dev\CleanupResult;
use Games\Paths\Core\Port\Dev\TestDataCleanupPort;
use PHPUnit\Framework\TestCase;
use Slim\Psr7\Factory\ResponseFactory;
use Slim\Psr7\Factory\ServerRequestFactory;

class DevControllerTest extends TestCase
{
    private function request()
    {
        return (new ServerRequestFactory())->createServerRequest('POST', '/api/dev/cleanup');
    }

    private function response()
    {
        return (new ResponseFactory())->createResponse();
    }

    public function testCleanupReturnsCountsWhenEnabled(): void
    {
        $port = $this->createMock(TestDataCleanupPort::class);
        $port->expects($this->once())->method('cleanupTestData')
            ->willReturn(new CleanupResult(5, 2));
        $controller = new DevController($port, true);

        $response = $controller->cleanup($this->request(), $this->response());

        $this->assertSame(200, $response->getStatusCode());
        $body = json_decode((string)$response->getBody(), true);
        $this->assertSame(['deletedGuests' => 5, 'deletedMatches' => 2], $body);
    }

    public function testCleanupReturns403WhenDisabled(): void
    {
        $port = $this->createMock(TestDataCleanupPort::class);
        $port->expects($this->never())->method('cleanupTestData');
        $controller = new DevController($port, false);

        $response = $controller->cleanup($this->request(), $this->response());

        $this->assertSame(403, $response->getStatusCode());
        $body = json_decode((string)$response->getBody(), true);
        $this->assertSame('DEV_ENDPOINTS_DISABLED', $body['error']);
    }
}
